<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="tensoRphylo">
<title>Input formats • tensoRphylo</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Input formats">
<meta property="og:description" content="tensoRphylo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tensoRphylo</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/tensoRphylo.html">Quickstart</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/modelingframework.html">Models</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/dataformat.html">Input</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/api.html">API</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-performance">Performance</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-performance">
    <a class="dropdown-item" href="../articles/validation.html">Validation</a>
    <a class="dropdown-item" href="../articles/comparison.html">Comparison</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Input formats</h1>
            
      
      
      <div class="d-none name"><code>dataformat.Rmd</code></div>
    </div>

    
    
<p>The input formats used by <code>tensorPhylo</code> are fairly
standard and intended to require minimal preprocessing: everything we do
is native to standard Newick, <code>NEXUS</code>, and
<code><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">ape::phylo</a></code> formats.</p>
<div class="section level2">
<h2 id="tree">Tree<a class="anchor" aria-label="anchor" href="#tree"></a>
</h2>
<p>The tree <strong>must</strong> be a rooted tree in
<code><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">ape::phylo</a></code> format. In general, <code><a href="https://rdrr.io/pkg/ape/man/read.nexus.html" class="external-link">ape::read.nexus</a></code>
does an excellent job of parsing valid Newick strings with the below
features such that they can be used in <code>tensoRphylo</code> without
any additional processing.</p>
<p>Trees <strong>may</strong> have any combination of extant and/or
extinct tips, as well as sampled ancestors. (In <code><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">ape::phylo</a></code>,
sampled ancestors are represented as nodes with a single
descendant.)</p>
<p>The tree <strong>may</strong> have a stem branch subtending the root,
in which case <code>tensoRphylo</code> will assume the process begins
with one lineage at the start of this stem branch. Otherwise,
<code>tensoRphylo</code> will assume the process begin with the two
lineages at the root of the tree. The length of the stem branch is
stored in the <code>root.edge</code> element of an
<code><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">ape::phylo</a></code> object.</p>
<div class="section level3">
<h3 id="a-closer-look-at-a-tree-with-extinct-tips-and-sampled-ancestors">A closer look at a tree with extinct tips and sampled ancestors<a class="anchor" aria-label="anchor" href="#a-closer-look-at-a-tree-with-extinct-tips-and-sampled-ancestors"></a>
</h3>
<p>The sampled-ancestor tree (with extant and/or extinct samples and
samples along branches) is the most complex type of tree produced by the
general modeling framework implemented in <code>tensoRphylo</code>; all
other trees (<em>e.g.</em>, trees of extant taxa only) are a special
case of this tree. You can find a nexus file with a valid Newick
representation of a sampled-ancestor tree in the package data (<a href="https://github.com/mikeryanmay/tensoRphylo/blob/main/inst/testdata/sampled_ancestor_tree.nex" class="external-link">link
to raw file</a>):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># path to tree file</span>
<span class="va">tree_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"sampled_ancestor_tree.nex"</span>,  
                         package <span class="op">=</span> <span class="st">"tensoRphylo"</span><span class="op">)</span>

<span class="co"># read the tree</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.nexus.html" class="external-link">read.nexus</a></span><span class="op">(</span><span class="va">tree_file</span><span class="op">)</span>

<span class="co"># print the tree</span>
<span class="va">tree</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## Phylogenetic tree with 14 tips and 18 internal nodes.</span>
<span class="co">## </span>
<span class="co">## Tip labels:</span>
<span class="co">##   t_70, t_71, t_17, t_72, t_73, t_66, ...</span>
<span class="co">## Node labels:</span>
<span class="co">##   , , , t_8, , t_16, ...</span>
<span class="co">## </span>
<span class="co">## Rooted; includes branch lengths.</span></code></pre>
<p>Notice that this tree has more internal nodes than tips, because some
of the nodes (the sampled ancestors) have only one descendant.</p>
<p>Inspecting the node labels, we see that some nodes are labeled
<code>""</code> and others are labeled with strings:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># print node labels</span>
<span class="va">tree</span><span class="op">$</span><span class="va">node.label</span></code></pre></div>
<pre><code><span class="co">##  [1] ""     ""     ""     "t_8"  ""     "t_16" ""     ""     "t_14" ""    </span>
<span class="co">## [11] ""     ""     ""     ""     ""     "t_53" ""     "t_81"</span></code></pre>
<p>The nodes with strings are the sampled ancestors; these node labels
are used to associate the sampled ancestor with character data
(discussed in the next section).</p>
<p>This tree also has a stem:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># print the root edge</span>
<span class="va">tree</span><span class="op">$</span><span class="va">root.edge</span></code></pre></div>
<pre><code><span class="co">## [1] 3.78442</span></code></pre>
<p>We can plot the tree to see that there are sampled ancestors.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot the tree with a root edge</span>
<span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></span><span class="op">(</span><span class="va">tree</span>, no.margin <span class="op">=</span> <span class="cn">TRUE</span>, root.edge <span class="op">=</span> <span class="cn">TRUE</span>, underscore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># plot the sampled ancestors with names</span>
<span class="va">sampled_ancestors</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">tree</span><span class="op">$</span><span class="va">Nnode</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/summary.phylo.html" class="external-link">Ntip</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">tree</span><span class="op">$</span><span class="va">node.label</span> <span class="op">!=</span> <span class="st">""</span><span class="op">]</span>
<span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span>node <span class="op">=</span> <span class="va">sampled_ancestors</span>, text <span class="op">=</span> <span class="va">tree</span><span class="op">$</span><span class="va">node.label</span><span class="op">[</span><span class="va">tree</span><span class="op">$</span><span class="va">node.label</span> <span class="op">!=</span> <span class="st">""</span><span class="op">]</span>, adj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, frame <span class="op">=</span> <span class="st">"none"</span>, srt <span class="op">=</span> <span class="fl">35</span><span class="op">)</span>
<span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span>node <span class="op">=</span> <span class="va">sampled_ancestors</span>, pch  <span class="op">=</span> <span class="fl">19</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="dataformat_files/figure-html/unnamed-chunk-4-1.png" alt="A tree with extant and extinct tips, as well as sampled ancestors (dots along branches)." width="95%"><p class="caption">
A tree with extant and extinct tips, as well as sampled ancestors (dots
along branches).
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="character-data">Character data<a class="anchor" aria-label="anchor" href="#character-data"></a>
</h2>
<p>The discrete character data can either be in the “traditional”
format, <em>i.e.</em>, a named vector specifying the state for each
sample, or a numeric matrix with rows corresponding to the samples and
the columns correspond to the states. I briefly describe each of these
formats below.</p>
<div class="section level3">
<h3 id="named-vector">Named vector<a class="anchor" aria-label="anchor" href="#named-vector"></a>
</h3>
<p>This format should be familiar to people who have used
<code>ape</code>, <code>phytools</code>, <code>diversitree</code>, etc.
In this format, the data are represented as a vector of numeric,
integer, or string variables, each element named by the corresponding
sample in the tree. While it is often convenient, the main drawback of
this format is that <code>tensoRphylo</code> must figure out the state
space of the model from the data. As a consequence,
<code>tensoRphylo</code> will only include states that are present in
the data. This will be inappropriate when you know there are states that
are not present in any sample; for example, if the state is the number
of chromosomes, then logically the state space should include all
integers, even if not every number of chromosomes is observed in the
data. I provide an example of a dataset with this format in the package
data:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># path to character data file</span>
<span class="va">data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"sampled_ancestor_data_vec.Rda"</span>,  
                         package <span class="op">=</span> <span class="st">"tensoRphylo"</span><span class="op">)</span>

<span class="co"># read the character data</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">data_file</span><span class="op">)</span>

<span class="va">data</span></code></pre></div>
<pre><code><span class="co">##  t_70  t_71  t_17  t_72  t_73  t_16   t_8  t_66  t_78  t_14  t_57  t_62  t_51 </span>
<span class="co">## "0/1"   "1"   "1"   "1"   "1"   "1"   "1"   "1"   "0"   "1"   "1"   "0"   "1" </span>
<span class="co">##  t_84  t_58  t_53  t_82  t_81  t_80 </span>
<span class="co">##   "1"   "1"   "1"   "1"   "1"   "1"</span></code></pre>
<p><code>tensoRphylo</code> has some <strong>added
functionality</strong> when provided with a <strong>named vector of
strings</strong>. First, it accommodates <em>missing</em> data,
represented as a <code>?</code> or <code>-</code>. Second, it
accommodates <em>ambiguous</em> data, <em>i.e.</em>, when a sample could
be one of many possible (perhaps hidden) states. Ambiguous data is
represented as a <code>/</code>-delimited string of the possible states,
as you see above. We can use the internal function
<code>tensoRphylo:::.char.vector.to.table</code> to see how
<code>tensoRphylo</code> translates this type of data into the numeric
matrix format, described next:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># make some missing data</span>
<span class="va">data</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"?"</span>
<span class="va">data</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"-"</span>

<span class="co"># convert to numeric matrix and print the result</span>
<span class="fu">tensoRphylo</span><span class="fu">:::</span><span class="fu">.char.vector.to.table</span><span class="op">(</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##      0 1</span>
<span class="co">## t_70 1 1</span>
<span class="co">## t_71 1 1</span>
<span class="co">## t_17 1 1</span>
<span class="co">## t_72 0 1</span>
<span class="co">## t_73 0 1</span>
<span class="co">## t_16 0 1</span>
<span class="co">## t_8  0 1</span>
<span class="co">## t_66 0 1</span>
<span class="co">## t_78 1 0</span>
<span class="co">## t_14 0 1</span>
<span class="co">## t_57 0 1</span>
<span class="co">## t_62 1 0</span>
<span class="co">## t_51 0 1</span>
<span class="co">## t_84 0 1</span>
<span class="co">## t_58 0 1</span>
<span class="co">## t_53 0 1</span>
<span class="co">## t_82 0 1</span>
<span class="co">## t_81 0 1</span>
<span class="co">## t_80 0 1</span></code></pre>
<p>Note that the top three rows (corresponding to the samples we set to
ambiguous/missing data) are now all <code>1</code>, because both states
are possible.</p>
</div>
<div class="section level3">
<h3 id="numeric-matrix">Numeric matrix<a class="anchor" aria-label="anchor" href="#numeric-matrix"></a>
</h3>
<p>In this case, the data are represented as a numeric matrix with one
row per sample and one column per state. The rows must be named
according to the corresponding sample in the tree; likewise, the columns
must be named according to the discrete state. The value of the
<em>i,j</em><sup>th</sup> element should be 1 if the
<em>i</em><sup>th</sup> sample <em>could be</em> state <em>j</em>, and 0
otherwise. I provide an example of a dataset with this format in the
package data:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># path to character data file</span>
<span class="va">data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"sampled_ancestor_data.Rda"</span>,  
                         package <span class="op">=</span> <span class="st">"tensoRphylo"</span><span class="op">)</span>

<span class="co"># read the character data</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">data_file</span><span class="op">)</span>

<span class="va">data</span></code></pre></div>
<pre><code><span class="co">##      0 1</span>
<span class="co">## t_70 1 1</span>
<span class="co">## t_71 0 1</span>
<span class="co">## t_17 0 1</span>
<span class="co">## t_72 0 1</span>
<span class="co">## t_73 0 1</span>
<span class="co">## t_16 0 1</span>
<span class="co">## t_8  0 1</span>
<span class="co">## t_66 0 1</span>
<span class="co">## t_78 1 0</span>
<span class="co">## t_14 0 1</span>
<span class="co">## t_57 0 1</span>
<span class="co">## t_62 1 0</span>
<span class="co">## t_51 0 1</span>
<span class="co">## t_84 0 1</span>
<span class="co">## t_58 0 1</span>
<span class="co">## t_53 0 1</span>
<span class="co">## t_82 0 1</span>
<span class="co">## t_81 0 1</span>
<span class="co">## t_80 0 1</span></code></pre>
</div>
<div class="section level3">
<h3 id="reading-character-data">Reading character data<a class="anchor" aria-label="anchor" href="#reading-character-data"></a>
</h3>
<div class="section level4">
<h4 id="nexus-format">
<code>NEXUS</code> format<a class="anchor" aria-label="anchor" href="#nexus-format"></a>
</h4>
<p><code>tensoRphylo</code> provides some basic functionality for
reading in character data in the desired format. The function
<code><a href="../reference/readNexusData.html">tensoRphylo::readNexusData</a></code> reads a <code>NEXUS</code> file
with a <em>single</em> discrete character (<code>nchar = 1</code>). This
function correctly interprets missing and ambiguous data, and determines
the number of states and the state labels from the <code>NEXUS</code>
file. Here we read in the sampled ancestor data we saw above (<a href="https://github.com/mikeryanmay/tensoRphylo/blob/main/inst/testdata/sampled_ancestor_data.nex" class="external-link">link
to raw file</a>):</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># path to character data file</span>
<span class="va">data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"sampled_ancestor_data.nex"</span>,  
                         package <span class="op">=</span> <span class="st">"tensoRphylo"</span><span class="op">)</span>

<span class="co"># read the character data</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readNexusData.html">readNexusData</a></span><span class="op">(</span><span class="va">data_file</span><span class="op">)</span>

<span class="va">data</span></code></pre></div>
<pre><code><span class="co">##      0 1</span>
<span class="co">## t_70 1 1</span>
<span class="co">## t_71 0 1</span>
<span class="co">## t_17 0 1</span>
<span class="co">## t_72 0 1</span>
<span class="co">## t_73 0 1</span>
<span class="co">## t_16 0 1</span>
<span class="co">## t_8  0 1</span>
<span class="co">## t_66 0 1</span>
<span class="co">## t_78 1 0</span>
<span class="co">## t_14 0 1</span>
<span class="co">## t_57 0 1</span>
<span class="co">## t_62 1 0</span>
<span class="co">## t_51 0 1</span>
<span class="co">## t_84 0 1</span>
<span class="co">## t_58 0 1</span>
<span class="co">## t_53 0 1</span>
<span class="co">## t_82 0 1</span>
<span class="co">## t_81 0 1</span>
<span class="co">## t_80 0 1</span></code></pre>
</div>
<div class="section level4">
<h4 id="delimited-text-format">Delimited text format<a class="anchor" aria-label="anchor" href="#delimited-text-format"></a>
</h4>
<p>The function <code><a href="../reference/readDelimitedData.html">tensoRphylo::readDelimitedData</a></code> reads a
comma- or tab-delimited file with one row per sample, the first column
containing the sample name, and the second column containing the
character data. The columns <em>should not</em> have headers. The
character data can including missing data (<code>?</code> or
<code>-</code>) or ambiguous data (a <code>/</code>-delimited string of
compatible states). The states should be encoded as numeric values
starting with 0. This format does not natively contain information about
the state space, so we must provide the number of states to the reader.
Again, we read in the sampled ancestor data we saw above (<a href="https://github.com/mikeryanmay/tensoRphylo/blob/main/inst/testdata/sampled_ancestor_data.csv" class="external-link">link
to raw file</a>).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># path to character data file</span>
<span class="va">data_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"testdata"</span>, <span class="st">"sampled_ancestor_data.csv"</span>,  
                         package <span class="op">=</span> <span class="st">"tensoRphylo"</span><span class="op">)</span>

<span class="co"># read the character data</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readDelimitedData.html">readDelimitedData</a></span><span class="op">(</span><span class="va">data_file</span>, delim <span class="op">=</span> <span class="st">","</span>, nstates <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>

<span class="va">data</span></code></pre></div>
<pre><code><span class="co">##      0 1</span>
<span class="co">## t_70 1 1</span>
<span class="co">## t_71 0 1</span>
<span class="co">## t_17 0 1</span>
<span class="co">## t_72 0 1</span>
<span class="co">## t_73 0 1</span>
<span class="co">## t_16 0 1</span>
<span class="co">## t_8  0 1</span>
<span class="co">## t_66 0 1</span>
<span class="co">## t_78 1 0</span>
<span class="co">## t_14 0 1</span>
<span class="co">## t_57 0 1</span>
<span class="co">## t_62 1 0</span>
<span class="co">## t_51 0 1</span>
<span class="co">## t_84 0 1</span>
<span class="co">## t_58 0 1</span>
<span class="co">## t_53 0 1</span>
<span class="co">## t_82 0 1</span>
<span class="co">## t_81 0 1</span>
<span class="co">## t_80 0 1</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="conformity">Conformity<a class="anchor" aria-label="anchor" href="#conformity"></a>
</h3>
<p><code>tensoRphylo</code> does some basic checks to make sure the tree
and data have the same samples. If you provide a data object with
samples that do not occur in the tree, <code>tensoRphylo</code> will
simply ignore those samples. Likewise, if you provide a tree with
samples that do not occur in the data, <code>tensoRphylo</code> will
simply add missing data for those samples to the provided dataset.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Michael R. May.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
